overlord/snapstate/snapstate.go

```diff
@@ -1978,10 +1992,94 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 	if len(updateTss.Refresh) > 0 {
 		updateTss.Refresh = finalizeUpdate(st, updateTss.Refresh, len(updates) > 0, updated, userID, flags)
 	}
-
 	return updated, updateTss, nil
 }
 
+// canSplitRefresh returns whether the refresh is a standard refresh of a mix
+// of essential and non-essential snaps on a hybrid system. If the refresh
+// can be split, it also returns the two split update groups.
+func canSplitRefresh(deviceCtx DeviceContext, infos []minimalInstallInfo, flags *Flags) (essential, nonEssential []minimalInstallInfo, split bool) {
+	// TODO: support split refresh for auto-refresh as well
+	if !deviceCtx.IsCoreBoot() || !release.OnClassic || flags.IsAutoRefresh {
+		return nil, nil, false
+	}
+
+	essential, nonEssential = splitEssentialUpdates(infos)
+	if len(essential) == 0 || len(nonEssential) == 0 {
+		return nil, nil, false
+	}
+
+	return essential, nonEssential, true
+}
+
+// splitRefresh creates two separate tasksets for the essential and non-essential
+// snap refresh groups, creating dependencies between specific tasks when
+// appropriate (e.g., snapd is present and should be refreshed first).
+func splitRefresh(essential, nonEssential []minimalInstallInfo, updateFunc func([]minimalInstallInfo) ([]string, *UpdateTaskSets, error)) ([]string, *UpdateTaskSets, error) {
+	// taskset with essential snaps (snapd, kernel and gadget)
+	essentialUpdated, essentialTss, err := updateFunc(essential)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// taskset with non-essential snaps (apps and their bases)
+	nonEssentialUpdated, nonEssentialTss, err := updateFunc(nonEssential)
```

Comment:
are we generating two rerefresh tasks now, one in each?

Reply:
sorry, I misread the code, that task is not added in doUpdate but one level up?

Reply:
Yes, it's added to the tasksets after the `doUpdate` call (that part was already like that before)